src/typescript-worker/types-fetcher.ts

```diff
@@ -9,244 +9,385 @@ import {ModuleResolver} from './module-resolver.js';
-      }
-      jsPath = packageJson.result.main ?? 'index.js';
-    }
-    const ext = fileExtension(jsPath);
-    if (ext === '') {
-      // No extension is presumed js.
-      jsPath += '.js';
-    } else if (ext !== 'js') {
-      // Unhandled kind of import.
-      return;
+    // Get the ".d.ts" path by changing extension, or looking up the "typings"
+    // field, etc.
+    location.path = await this._getDtsPath(location);
+    // Resolve the concrete version.
+    if (!handledByImportMap) {
+      location = await this._cdn.canonicalize(location);
     }
-    const dtsPath = changeFileExtension(jsPath, 'd.ts');
-    const dtsSpecifier = `${pkg}/${dtsPath}`;
-    if (this._handledSpecifiers.has(dtsSpecifier)) {
+    if (referrer === root || location.pkg !== referrer.pkg) {
+      // Note the two package names can be the same in the case that a typings
+      // file imports another module from the same package using its own bare
+      // module name, instead of a relative path. TypeScript does support this
+      // case, and it does happen (e.g. @material/base does this). We don't need
+      // a self-edge in the dependency graph, though.
+      this._addEdgeToDependencyGraph(referrer, location);
+    }
+    // Stop early if we've already handled this specifier.
+    if (
+      this._filesByPackageVersion
+        .get(location.pkg)
+        ?.get(location.version)
+        ?.get(location.path) !== undefined
+    ) {
       return;
     }
-    this._handledSpecifiers.add(dtsSpecifier);
-    const dtsResult = await this._fetchAsset(dtsSpecifier);
+    // Ready to fetch and recurse.
+    const dtsResult = await this._fetchAndAddToOutputFiles(location);
     if (dtsResult.error !== undefined) {
       return;
     }
-    await this._handleBareAndRelativeSpecifiers(dtsResult.result, {
-      pkg,
-      path: jsPath,
-    });
-  }
-
-  private _resolveBareModule(bare: string): string {
-    const result = this._moduleResolver.resolve(bare, '');
-    if (result.type !== 'bare') {
-      throw new Error(`Only expected bare module for specifier ${bare}`);
-    }
-    return result.url;
+    await this._handleBareAndRelativeSpecifiers(dtsResult.result, location);
   }
 
   private async _handleRelativeSpecifier(
     relative: string,
-    referrerSpecifier: NodePackage
+    referrer: NpmFileLocation
   ): Promise<void> {
-    const ext = fileExtension(relative);
-    if (ext === '') {
-      // No extension is presumed js.
-      relative += '.js';
-    } else if (ext !== 'js') {
-      // Unhandled kind of import.
+    const location = {
+      // We know package and version must be the same as the referrer, since
+      // this is a relative path, and we are therefore still in the same
+      // package.
+      pkg: referrer.pkg,
+      version: referrer.version,
+      // Make the path package-root relative, instead of referrer-path relative.
+      path: resolveUrlPath(referrer.path, relative).slice(1), // Remove the leading '/'.
```

Comment:
Nit: Rather than the comment is it worth using the `trimLeadingSlash` util.

Reply:
Good call, done!